mm2src/coins/eth/nft_swap_v2/mod.rs

```diff
@@ -305,6 +367,80 @@ impl EthCoin {
         Ok(data)
     }
 
+    fn prepare_refund_nft_maker_payment_v2_timelock(
+        &self,
+        args: &RefundNftMakerPaymentArgs<'_, Self>,
+        decoded: Vec<Token>,
+        htlc_params: Vec<Token>,
+        state: U256,
+    ) -> Result<Vec<u8>, PrepareTxDataError> {
+        validate_state(args.maker_payment_tx, state, MakerPaymentStateV2::PaymentSent as u8)?;
+
+        let refund_func = match args.contract_type {
+            ContractType::Erc1155 => NFT_MAKER_SWAP_V2.function("refundErc1155MakerPaymentTimelock")?,
+            ContractType::Erc721 => NFT_MAKER_SWAP_V2.function("refundErc721MakerPaymentTimelock")?,
+        };
+
+        let input_tokens = match args.contract_type {
+            ContractType::Erc1155 => vec![
+                htlc_params[0].clone(), // swapId
+                htlc_params[1].clone(), // takerAddress
+                Token::FixedBytes(args.taker_secret_hash.to_vec()),
+                Token::FixedBytes(args.maker_secret_hash.to_vec()),
+                htlc_params[2].clone(), // tokenAddress
+                decoded[2].clone(),     // tokenId
+                decoded[3].clone(),     // amount
+            ],
+            ContractType::Erc721 => vec![
+                htlc_params[0].clone(), // swapId
+                htlc_params[1].clone(), // takerAddress
+                Token::FixedBytes(args.taker_secret_hash.to_vec()),
+                Token::FixedBytes(args.maker_secret_hash.to_vec()),
+                htlc_params[2].clone(), // tokenAddress
+                decoded[2].clone(),     // tokenId
+            ],
+        };
+        let data = refund_func.encode_input(&input_tokens)?;
+        Ok(data)
```

Comment:
do you really need to clone here? what about this
```rust 
    let input_tokens = match args.contract_type {
            ContractType::Erc1155 => vec![
                &htlc_params[0], // swapId
                &htlc_params[1], // takerAddress
                &Token::FixedBytes(args.taker_secret_hash.to_vec()),
                &Token::FixedBytes(args.maker_secret_hash.to_vec()),
                &htlc_params[2], // tokenAddress
                &decoded[2],     // tokenId
                &decoded[3],     // amount
            ],
            ContractType::Erc721 => vec![
                &htlc_params[0], // swapId
                &htlc_params[1], // takerAddress
                Token::FixedBytes(args.taker_secret_hash.to_vec()),
                Token::FixedBytes(args.maker_secret_hash.to_vec()),
                &htlc_params[2], // tokenAddress
                &decoded[2],     // tokenId
            ],
        };
```

Reply:
The suggestion above will result in a `Vec<&Token>` type, but we need to provide a `&[Token]` to `encode_input()`. So, we need to build a new `Vector<Token>` from the `htlc_params` vector and `args` fields. That's why we have to use `clone` and then pass `Vec<Token>` as a slice.

Ive noticed some code duplication in `input_tokens` creation, so I did this https://github.com/KomodoPlatform/komodo-defi-framework/pull/2129/commits/e636da1e5603f3253adf1d79908f266f4fcb3e8f